StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c

```diff
@@ -25,6 +25,203 @@ EFI_MM_COMMUNICATION_PROTOCOL  mMmCommunication = {
 
 MM_COMM_BUFFER             mMmCommonBuffer;
 EFI_SMM_CONTROL2_PROTOCOL  *mSmmControl2;
+EFI_SMM_ACCESS2_PROTOCOL   *mSmmAccess;
+BOOLEAN                    mSmmLocked = FALSE;
+BOOLEAN                    mEndOfDxe  = FALSE;
+
+//
+// Data structure used to declare a table of protocol notifications and event
+// notifications required by the Standalone Mm environment
+//
+typedef struct {
+  BOOLEAN             Protocol;
+  BOOLEAN             CloseOnLock;
+  EFI_GUID            *Guid;
+  EFI_EVENT_NOTIFY    NotifyFunction;
+  VOID                *NotifyContext;
+  EFI_TPL             NotifyTpl;
+  EFI_EVENT           Event;
+} MM_EVENT_NOTIFICATION;
+
+//
+// Table of Protocol notification and GUIDed Event notifications that the Standalone Mm requires
+//
+MM_EVENT_NOTIFICATION  mMmEvents[] = {
+  //
+  // Declare protocol notification on DxeSmmReadyToLock protocols.  When this notification is established,
+  // the associated event is immediately signalled, so the notification function will be executed and the
+  // DXE SMM Ready To Lock Protocol will be found if it is already in the handle database.
+  //
+  { TRUE,  TRUE,  &gEfiDxeSmmReadyToLockProtocolGuid, MmReadyToLockEventNotify,     &gEfiDxeSmmReadyToLockProtocolGuid, TPL_CALLBACK, NULL },
+  //
+  // Declare event notification on EndOfDxe event.  When this notification is established,
+  // the associated event is immediately signalled, so the notification function will be executed and the
+  // SMM End Of Dxe Protocol will be found if it is already in the handle database.
+  //
+  { FALSE, TRUE,  &gEfiEndOfDxeEventGroupGuid,        MmGuidedEventNotify,          &gEfiEndOfDxeEventGroupGuid,        TPL_CALLBACK, NULL },
+  //
+  // Declare event notification on EndOfDxe event.  This is used to set EndOfDxe event signaled flag.
+  //
+  { FALSE, TRUE,  &gEfiEndOfDxeEventGroupGuid,        MmEndOfDxeEventNotify,        &gEfiEndOfDxeEventGroupGuid,        TPL_CALLBACK, NULL },
+  //
+  // Declare event notification on Ready To Boot Event Group.  This is an extra event notification that is
+  // used to make sure SMRAM is locked before any boot options are processed.
+  //
+  { FALSE, TRUE,  &gEfiEventReadyToBootGuid,          MmReadyToLockEventNotify,     &gEfiEventReadyToBootGuid,          TPL_CALLBACK, NULL },

```

Comment:
Do you think that moving this to just below "gEfiDxeSmmReadyToLockProtocolGuid" entry is more readable?

Reply:
Moved it to just below "gEfiDxeSmmReadyToLockProtocolGuid" entry